### DISCUSSION TITLE
Error: Failed to prepare input devices: No device was selected!

### DISCUSSION THREAD

**Discussion by EricDriussi** [xremap/xremap#436] [Q&A] [answered]
Hi there!
I successfully installed xremap on my arch install but for some reason I'm not able to get it going on my fedora install.

I downloaded the binary from the release page as per the readme, and ran the following command to get it to work without sudo:

```
sudo gpasswd -a YOUR_USER input
echo 'KERNEL=="uinput", GROUP="input", TAG+="uaccess"' | sudo tee /etc/udev/rules.d/input.rules
```

No matter if I run it with or without sudo, I always get this message: 

```
Selecting devices from the following list:
------------------------------------------------------------------------------
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
Error: Failed to prepare input devices: No device was selected!
```

`ls -la /dev/input/by-id/` produces this output:

```
lrwxrwxrwx. 1 root root  10 Mar 14 16:00 usb-1ea7_2.4G_Mouse-event-if01 -> ../event15
lrwxrwxrwx. 1 root root  10 Mar 14 16:00 usb-1ea7_2.4G_Mouse-event-kbd -> ../event12
lrwxrwxrwx. 1 root root  10 Mar 14 16:00 usb-1ea7_2.4G_Mouse-if01-event-mouse -> ../event14
lrwxrwxrwx. 1 root root   9 Mar 14 16:00 usb-1ea7_2.4G_Mouse-if01-mouse -> ../mouse4
lrwxrwxrwx. 1 root root  10 Mar 14 16:00 usb-Logitech_G733_Gaming_Headset-event-if03 -> ../event20
lrwxrwxrwx. 1 root root   9 Mar 14 16:00 usb-ZSA_Technology_Labs_Moonlander_Mark_I-event-if03 -> ../event9
lrwxrwxrwx. 1 root root   9 Mar 14 16:00 usb-ZSA_Technology_Labs_Moonlander_Mark_I-event-kbd -> ../event7
lrwxrwxrwx. 1 root root   9 Mar 14 16:00 usb-ZSA_Technology_Labs_Moonlander_Mark_I-if02-event-mouse -> ../event8
lrwxrwxrwx. 1 root root   9 Mar 14 16:00 usb-ZSA_Technology_Labs_Moonlander_Mark_I-if02-mouse -> ../mouse3
lrwxrwxrwx. 1 root root  10 Mar 14 16:00 usb-ZSA_Technology_Labs_Moonlander_Mark_I-if03-event-kbd -> ../event11
```

Following these same steps produced no errors on arch, I also tried installing with cargo but I've had no luck on Fedora.
Any idea what I'm missing?

**Comment by k0kubun** [comment]
That's weird. I'd expect it to work at least when you use sudo.

> ls -la /dev/input/by-id/ produces this output:

I see that it's symlinked from `/dev/input`, but could you also show the output of `ls -la /dev/input`? (Please show an entire output of the command when you do so, including things like `.`)

**Reply by EricDriussi** [reply]
Sure! `ls -la /dev/input` looks like this:

```
drwxr-xr-x.  4 root root     800 Mar 15 08:08 .
drwxr-xr-x. 20 root root    4620 Mar 15 08:08 ..
drwxr-xr-x.  2 root root     240 Mar 15 08:08 by-id
drwxr-xr-x.  2 root root     600 Mar 15 08:08 by-path
crw-rw----.  1 root input 13, 64 Mar 15 08:08 event0
crw-rw----.  1 root input 13, 65 Mar 15 08:08 event1
crw-rw----.  1 root input 13, 74 Mar 15 08:08 event10
crw-rw----.  1 root input 13, 75 Mar 15 08:08 event11
crw-rw----.  1 root input 13, 76 Mar 15 08:08 event12
crw-rw----.  1 root input 13, 77 Mar 15 08:08 event13
crw-rw----.  1 root input 13, 78 Mar 15 08:08 event14
crw-rw----.  1 root input 13, 79 Mar 15 08:08 event15
crw-rw----.  1 root input 13, 80 Mar 15 08:08 event16
crw-rw----.  1 root input 13, 81 Mar 15 08:08 event17
crw-rw----.  1 root input 13, 82 Mar 15 08:08 event18
crw-rw----.  1 root input 13, 83 Mar 15 08:08 event19
crw-rw----.  1 root input 13, 66 Mar 15 08:08 event2
crw-rw----.  1 root input 13, 84 Mar 15 08:08 event20
crw-rw----.  1 root input 13, 85 Mar 15 08:08 event21
crw-rw----.  1 root input 13, 86 Mar 15 08:08 event22
crw-rw----.  1 root input 13, 87 Mar 15 08:08 event23
crw-rw----.  1 root input 13, 88 Mar 15 08:08 event24
crw-rw----.  1 root input 13, 89 Mar 15 08:08 event25
crw-rw----.  1 root input 13, 90 Mar 15 08:08 event26
crw-rw----.  1 root input 13, 91 Mar 15 08:08 event27
crw-rw----.  1 root input 13, 92 Mar 15 08:08 event28
crw-rw----.  1 root input 13, 93 Mar 15 08:08 event29
crw-rw----.  1 root input 13, 67 Mar 15 08:08 event3
crw-rw----.  1 root input 13, 68 Mar 15 08:08 event4
crw-rw----.  1 root input 13, 69 Mar 15 08:08 event5
crw-rw----.  1 root input 13, 70 Mar 15 08:08 event6
crw-rw----.  1 root input 13, 71 Mar 15 08:08 event7
crw-rw----.  1 root input 13, 72 Mar 15 08:08 event8
crw-rw----.  1 root input 13, 73 Mar 15 08:08 event9
crw-rw----.  1 root input 13, 63 Mar 15 08:08 mice
crw-rw----.  1 root input 13, 32 Mar 15 08:08 mouse0
crw-rw----.  1 root input 13, 33 Mar 15 08:08 mouse1
crw-rw----.  1 root input 13, 34 Mar 15 08:08 mouse2
crw-rw----.  1 root input 13, 35 Mar 15 08:08 mouse3
crw-rw----.  1 root input 13, 36 Mar 15 08:08 mouse4
```

**Reply by EricDriussi** [reply]
Oh and by the way, just in case it helps: both my Arch and Fedora install are on different laptops, both using the same input devices through a KVM (well apart from the integrated ones of course).

**Reply by EricDriussi** [reply, marked as answer]
Well, for some reason and after a couple of re-installs/re-compiles, `cargo install xremap --features x11` seems to have worked now o.O
Not really sure what happened there (probably some human error on my side) but it's all good now!

**Reply by El-Barto89** [reply]
hi all, same problem, I can't figure out what's wrong. Can I ask for help? Thank you in advance.

-Ubuntu 22.04
Installed autokey, keyd, solaar.

my output of ls -la /dev/input is:
drwxr-xr-x  2 root root     360 mag  1 19:42 .
drwxr-xr-x 20 root root    5040 mag  1 19:42 ..
crw-rw----  1 root input 13, 64 mag  1 19:02 event0
crw-rw----  1 root input 13, 65 mag  1 19:02 event1
crw-rw----  1 root input 13, 74 mag  1 19:03 event10
crw-rw----  1 root input 13, 75 mag  1 19:42 event11
crw-rw----  1 root input 13, 66 mag  1 19:02 event2
crw-rw----  1 root input 13, 67 mag  1 19:03 event3
crw-rw----  1 root input 13, 68 mag  1 19:03 event4
crw-rw----  1 root input 13, 69 mag  1 19:03 event5
crw-rw----  1 root input 13, 70 mag  1 19:03 event6
crw-rw----  1 root input 13, 71 mag  1 19:03 event7
crw-rw----  1 root input 13, 72 mag  1 19:03 event8
crw-rw----  1 root input 13, 73 mag  1 19:03 event9
crw-rw-r--  1 root input 13,  0 mag  1 19:03 js0
crw-rw----  1 root input 13, 63 mag  1 19:02 mice
crw-rw----  1 root input 13, 32 mag  1 19:03 mouse0
crw-rw----  1 root input 13, 33 mag  1 19:42 mouse1

**Reply by k0kubun** [reply]
Do you get the same issue even if you're running xremap with sudo?

You might need udev rules and your user to be in `input` group, so please also share the output of commands that show those configurations.

**Reply by El-Barto89** [reply]
yes, same problem also with sudo xremap. Actually, if I launch xremap with sudo this is my output:
`sudo: xremap: command not found
`
 but I installed rust/cargo correctly (at least i think so)
`cargo --version →  cargo 1.77.2 (e52e36006 2024-03-26)
`
`xremap --version → xremap 0.8.18
`
Can I ask for the correct commands to show you correct output? I'm relative new to Linux. (I think I'm already into input group because I've already launched `sudo usermod -a -G input myusername` and If I check into /etc/group I see: `input:x:107:myusername`

Thank you for your precious time

**Reply by k0kubun** [reply]
Can you find the full path of xremap using `which xremap` and then run that full path with `sudo`?

**Comment by El-Barto89** [comment]
> Can you find the full path of xremap using `which xremap` and then run that full path with `sudo`?

OK, I can confirm that somehow the problem is related to [KEYD](https://github.com/rvaiya/keyd), I disabled it and run sudo with the full path that you suggested to me. It finally works, the only thing is that it runs only while my terminal is open, not in the background. Is this its normal behavior?

**Reply by k0kubun** [reply]
It's the expected behavior. I manage a xremap process using systemd and auto-start the service using a desktop entry.

Here's my [xremap.service](https://github.com/k0kubun/dotfiles/blob/7c423b2f664867498990d8df21a2e29cccdc5bf3/recipes/ubuntu/files/home/k0kubun/.config/systemd/user/xremap.service) (`ExecStartPre` is not necessary, and your `ExecStart` path would be different) and [xremap.desktop](https://github.com/k0kubun/dotfiles/blob/7c423b2f664867498990d8df21a2e29cccdc5bf3/recipes/ubuntu/files/home/k0kubun/.config/autostart/xremap.desktop).

**Comment by 12risobo** [comment]
I tried to install Xremap for Wayland Ubuntu 

First I installed Rust. 
Then 
`cargo install xremap --features gnome`


`sudo xremap  ~/.config/xremap/config.yml
sudo: xremap: command not found`

`xremap  ~/.config/xremap/config.yml`
gives me the following error:

Selecting devices from the following list:
------------------------------------------------------------------------------
------------------------------------------------------------------------------
Selected keyboards automatically since --device options weren't specified:
------------------------------------------------------------------------------
Error: Failed to prepare input devices: No device was selected!

My output of ls -la /dev/input is:
total 0
drwxr-xr-x  4 root root     500 nov  1 18:29 .
drwxr-xr-x 21 root root    5520 nov  1 18:29 ..
drwxr-xr-x  2 root root     120 nov  1 18:29 by-id
drwxr-xr-x  2 root root     280 nov  1 18:29 by-path
crw-rw----  1 root input 13, 64 nov  1 18:29 event0
crw-rw----  1 root input 13, 65 nov  1 18:29 event1
crw-rw----  1 root input 13, 74 nov  1 18:29 event10
crw-rw----  1 root input 13, 75 nov  1 18:29 event11
crw-rw----  1 root input 13, 76 nov  1 18:29 event12
crw-rw----  1 root input 13, 77 nov  1 18:29 event13
crw-rw----  1 root input 13, 78 nov  1 18:29 event14
crw-rw----  1 root input 13, 79 nov  1 18:29 event15
crw-rw----  1 root input 13, 80 nov  1 18:29 event16
crw-rw----  1 root input 13, 66 nov  1 18:29 event2
crw-rw----  1 root input 13, 67 nov  1 18:29 event3
crw-rw----  1 root input 13, 68 nov  1 18:29 event4
crw-rw----  1 root input 13, 69 nov  1 18:29 event5
crw-rw----  1 root input 13, 70 nov  1 18:29 event6
crw-rw----  1 root input 13, 71 nov  1 18:29 event7
crw-rw----  1 root input 13, 72 nov  1 18:29 event8
crw-rw----  1 root input 13, 73 nov  1 18:29 event9
crw-rw----  1 root input 13, 63 nov  1 18:29 mice
crw-rw----  1 root input 13, 32 nov  1 18:29 mouse0
crw-rw----  1 root input 13, 33 nov  1 18:29 mouse1
crw-rw----  1 root input 13, 34 nov  1 18:29 mouse2

I Installed the GNOME Shell extension:
![Xremap Gnome Shell Extensions](https://github.com/user-attachments/assets/0d7aec69-0ba5-4527-a993-94034bdb2530)

What else can I try? I am rather new at Linux Ubuntu and am open to any suggestions.
